### PR TITLE
Kf 2268 feat tensorflow full rock

### PR DESCRIPTION
**Pull request by i-chvets**
Tensorflow full initial ROCK.

Summary of changes:
- Initial ROCK version with latest updates and create according to best practices spec.
- Tests for testing ROCK integrity.
- Commented out tests as TO-DOs for next step of ROCK tests that will serve as framework for all other tests in Notebook servers.

Testing was done using tests setup for this ROCK in this PR. This testing follows manual testing procedure describe in best practices spec.

To test ROCK:
```
tox -e unit
```

### CODE REVIEW

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+services:
+  jupyter:
+    override: replace
+    summary: "jupyter-tensorflow-full service"
+    startup: enabled
+    environment:
+      NB_USER: jovyan
+      NB_UID: 1001
+      NB_PREFIX: /
+      HOME: /home/jovyan
```
Comment (SamirPS) [inline jupyter-tensorflow-full/rockcraft.yaml:23]: same

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+      SHELL: /bin/bash
+      LANG: en_US.UTF-8
+      LANGUAGE: en_US.UTF-8
+      LC_ALL: en_US.UTF-8
+      PATH: /opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+    command: /opt/conda/bin/jupyter lab --notebook-dir="/home/jovyan" --ip=0.0.0.0 --no-browser --port=8888 --ServerApp.token="" --ServerApp.password="" --ServerApp.allow_origin="*" --ServerApp.base_url="/" --ServerApp.authenticate_prometheus=False
```
Comment (SamirPS) [inline jupyter-tensorflow-full/rockcraft.yaml:29]: use working-dir

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    source-tag: v1.7-branch # upstream branch
+    build-packages:
+      - lsb-release
+      - wget
+    build-environment:
+      - MINIFORGE_ARCH: "x86_64"
+      - MINIFORGE_VERSION: "4.10.1-4"
+      - PIP_VERSION: "21.1.2"
+      - PYTHON_VERSION: "3.8.10"
+      - HOME: "/home/jovyan"
```
Comment (SamirPS) [inline jupyter-tensorflow-full/rockcraft.yaml:108]: why jovyan ?

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+      # Download the miniforge install shasums
+      wget -qO miniforge.sh.sha256 \
+        "https://github.com/conda-forge/miniforge/releases/download/${MINIFORGE_VERSION}/Miniforge3-${MINIFORGE_VERSION}-Linux-${MINIFORGE_ARCH}.sh.sha256"
+
+      # Verify that the downloaded installer is correct
+      echo "$(cat miniforge.sh.sha256 | awk '{ print $1; }') miniforge.sh" | sha256sum --check
+
+      # Run the miniforge installer
+      /bin/sh miniforge.sh -b -f -p "${CONDA_DIR}"
```
Comment (SamirPS) [inline jupyter-tensorflow-full/rockcraft.yaml:127]: by default it's /bin/sh https://snapcraft.io/docs/snapcraft-parts-metadata#heading--override-build

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+      # Configure conda
+      "${CONDA_BIN}/conda" config --system --set auto_update_conda false
+      "${CONDA_BIN}/conda" config --system --set show_channel_urls true
+
+      echo "conda ${MINIFORGE_VERSION:0:-2}" >> "${CONDA_DIR}/conda-meta/pinned"
+      echo "python ${PYTHON_VERSION}" >> "${CONDA_DIR}/conda-meta/pinned"
+
+      # Install the correct versions of python, conda, pip
+      "${CONDA_BIN}/conda" install -y -q \
```
Comment (SamirPS) [inline jupyter-tensorflow-full/rockcraft.yaml:137]: Can you put python as a build-packages requirements ?

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+      # Clean up
+      "${CONDA_BIN}/conda" update -y -q --all
+      "${CONDA_BIN}/conda" clean -a -f -y
+
+      # Install the jupyter python requirements
+      cp ${CRAFT_PART_SRC}/components/example-notebook-servers/jupyter/requirements.txt jupyter-requirements.txt
+      "${CONDA_BIN}/pip" install --no-cache-dir -r jupyter-requirements.txt
+
+      # Generate a jupyter lab config
+      "${CONDA_BIN}/jupyter" lab --generate-config --allow-root
```
Comment (SamirPS) [inline jupyter-tensorflow-full/rockcraft.yaml:151]: why do you need root ? ( i don't know a lot about conda)

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+      echo "conda activate base" >> "${CRAFT_PART_INSTALL}/etc/profile"
+
+      # Do some cleanup
+      rm -rf "${HOME}/.wget-hsts" "${HOME}/.cache"
+
+      # Stage the conda directories
+      cp -r "${CONDA_DIR}" "${CRAFT_PART_INSTALL}/opt/conda"
+      cp -r "${HOME}" "${CRAFT_PART_INSTALL}/${HOME}"
+
+  # not-root user for this ROCK should be 'jovyan'
```
Comment (SamirPS) [inline jupyter-tensorflow-full/rockcraft.yaml:177]: why jovyan ?

**Review by SamirPS — CHANGES_REQUESTED** [review]

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+license: Apache-2.0
+base: ubuntu:22.04
+run-user: _daemon_
+services:
+  jupyter:
+    override: replace
+    summary: "jupyter-tensorflow-full service"
+    startup: enabled
+    environment:
+      NB_USER: jovyan
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:20]: Required by notebook server.

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 1 above)
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:23]: Required by notebook server.

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 3 above)
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:108]: Required by notebook server.

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 7 above)
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:177]: Required by notebook server.

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 2 above)
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:29]: Will do.

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+  kubectl:
+    plugin: nil
+    stage-snaps:
+      - kubectl/1.24/stable
+    organize:
+      kubectl: bin/kubectl
+    stage:
+      - bin/kubectl
+
+  nodejs:
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:80]: Coming from [here](https://github.com/kubeflow/kubeflow/blob/master/components/example-notebook-servers/jupyter/Dockerfile)

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 5 above)
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:137]: This is also based on original [Dockerfile](https://github.com/kubeflow/kubeflow/blob/v1.7-branch/components/example-notebook-servers/jupyter/Dockerfile)

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 6 above)
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:151]: Need to verify.

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+      SHELL: /bin/bash
+      LANG: en_US.UTF-8
+      LANGUAGE: en_US.UTF-8
+      LC_ALL: en_US.UTF-8
+      PATH: /opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+    command: /opt/conda/bin/jupyter lab --notebook-dir="/home/jovyan" --ip=0.0.0.0 --no-browser --port=8888 --ServerApp.token="" --ServerApp.password="" --ServerApp.allow_origin="*" --ServerApp.base_url="/" --ServerApp.authenticate_prometheus=False
+platforms:
+  amd64:
+
+package-repositories:
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:33]: It comes from [here](https://github.com/kubeflow/kubeflow/blob/v1.7-branch/components/example-notebook-servers/jupyter/Dockerfile#L17)

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 6 above)
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:151]: Will remove.

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 5 above)
Comment (SamirPS) [reply jupyter-tensorflow-full/rockcraft.yaml:137]: But you need Python only for the build, no? so why not install using build-packages?

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+      (echo "# os-release" && cat /etc/os-release && echo "# dpkg-query" && \
+      dpkg-query -f '${db:Status-Abbrev},${binary:Package},${Version},${source:Package},${Source:Version}\n' -W) \
+      > ${CRAFT_PART_INSTALL}/usr/share/rocks/dpkg.query
+
+  overlay-pkgs:
+    plugin: nil
+    # These were taken from the upstream base image. They *may not* all be required.
+    overlay-packages:
+      - apt-transport-https
+      - bash
```
Comment (SamirPS) [inline jupyter-tensorflow-full/rockcraft.yaml:53]: i think you can remove bash and other thing

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 4 above)
Comment (SamirPS) [reply jupyter-tensorflow-full/rockcraft.yaml:128]: you can remove ``/bin/sh`` normally

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+    build-environment:
+      - MINIFORGE_ARCH: "x86_64"
+      - MINIFORGE_VERSION: "4.10.1-4"
+      - PIP_VERSION: "21.1.2"
+      - PYTHON_VERSION: "3.8.10"
+      - HOME: "/home/jovyan"
+      - CONDA_DIR: "/opt/conda"
+      - CONDA_BIN: "/opt/conda/bin"
+    stage-packages:
+      - python3.8-distutils
```
Comment (SamirPS) [inline jupyter-tensorflow-full/rockcraft.yaml:111]: it's conda which install python on build time ?

**Review by SamirPS — CHANGES_REQUESTED** [review]

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 11 above)
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:53]: I would like to have bash availbale for users.

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 12 above)
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:111]: This one is missing for some reason. Had to add it this way.

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 11 above)
Comment (SamirPS) [reply jupyter-tensorflow-full/rockcraft.yaml:52]: and i don"t see also the reason behind have bash for user

File: jupyter-tensorflow-full/rockcraft.yaml [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+    plugin: nil
+    override-build: |
+      mkdir -p ${CRAFT_PART_INSTALL}/usr/share/rocks
+      (echo "# os-release" && cat /etc/os-release && echo "# dpkg-query" && \
+      dpkg-query -f '${db:Status-Abbrev},${binary:Package},${Version},${source:Package},${Source:Version}\n' -W) \
+      > ${CRAFT_PART_INSTALL}/usr/share/rocks/dpkg.query
+
+  overlay-pkgs:
+    plugin: nil
+    overlay-packages:
```
Comment (SamirPS) [inline jupyter-tensorflow-full/rockcraft.yaml:50]: You use base:ubuntu:20.04 so bash, bzip2 for example will be on the ROCK. Also i don"t see also the reason behind have bash for user

**Review by SamirPS — CHANGES_REQUESTED** [review]

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 13 above)
Comment (cjdcordeiro) [inline jupyter-tensorflow-full/rockcraft.yaml:50]: leaving my 2 cents here:
 - beside what @SamirPS already mentioned (wrt many of these utilities already being in the base image), I think you should reconsider this ROCK's purpose. In a simplistic way, you can usually find 3 types of container images: application images (like nginx, prometheus, etc.), distribution images (like ubuntu, alpine), and runtime images (like python, golang). ROCKs are, typically, application images, which means it is weird to find this amount of utilities within. You might have a good reason for having things like `bash`, I won't dispute that, but OTOH, I'd question the inclusion of things like `curl`, `git`, `vim`. Why would these be necessary for an application image? If your answer is: "my users need to be able to debug the image and interact with it", then I'd say: "they can already do that cause your ROCK is based on `ubuntu` and thus already has all the essentials for that. If someone wants an extra utility, like `vim`, they can just `apt install vim`".

Anyway, I might be missing context, but I just wanted to share my view on this cause I feel the resulting ROCK is getting unnecessarily big, complex and prone to vulnerabilities.

**Review by cjdcordeiro — COMMENTED** [review]
:eyes: just leaving a quick comment :)

File: jupyter-tensorflow-full/rockcraft.yaml (hunk 13 above)
Comment (i-chvets) [reply jupyter-tensorflow-full/rockcraft.yaml:50]: @cjdcordeiro Thank you for the input. Our Kubeflow ROCKs are strange beasts.
This particular set of images are servers that users interact with and run their own Python scripts with. These are not really an application running by itself, but rather used as base for data scientists to run their own workloads. End users would defintely use `git` and `curl`, `vim` not so much. Some packages could be considered extras, but we would like to keep same user experience as upstream, that's the reason we are keeping all those packages in. We would like end users to not install something that they are used to having in upstream image.

Comment (cjdcordeiro) [reply jupyter-tensorflow-full/rockcraft.yaml:50]: ack. Anyway, whenever we have the chance to improve wrt upstream, we should do it (in this case, stuff like `vim` is an obvious candidate to be dropped)

**Review by ca-scribner — APPROVED** [review]

**Pull request merged**